pkg/customresourcestate/generate/generator/generator.go

```diff
@@ -0,0 +1,140 @@
+)
+
+// CustomResourceConfigGenerator implements the Generator interface from controller-tools.
+// It uses markers to generate a custom resource configuration for kube-state-metrics from go code.
+type CustomResourceConfigGenerator struct{}
+
+var _ genall.Generator = &CustomResourceConfigGenerator{}
+var _ genall.NeedsTypeChecking = &CustomResourceConfigGenerator{}
+
+// RegisterMarkers registers all markers needed by this Generator
+// into the given registry.
+func (g CustomResourceConfigGenerator) RegisterMarkers(into *ctrlmarkers.Registry) error {
+	for _, m := range markers.MarkerDefinitions {
+		if err := m.Register(into); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Generate generates artifacts produced by this marker.
+// It's called after RegisterMarkers has been called.
+func (g CustomResourceConfigGenerator) Generate(ctx *genall.GenerationContext) error {
+	// Create the parser which is specific to the metric generator.
+	parser := newParser(
+		&crd.Parser{
+			Collector: ctx.Collector,
+			Checker:   ctx.Checker,
+		},
+	)
+
+	// Loop over all passed packages.
+	for _, pkg := range ctx.Roots {
+		// skip packages which don't import metav1 because they can't define a CRD without meta v1.
+		metav1 := pkg.Imports()["k8s.io/apimachinery/pkg/apis/meta/v1"]
+		if metav1 == nil {
+			continue
+		}
+
+		// parse the given package to feed crd.FindKubeKinds with Kubernetes Objects.
+		parser.NeedPackage(pkg)
+
+		kubeKinds := crd.FindKubeKinds(parser.Parser, metav1)
+		if len(kubeKinds) == 0 {
+			klog.Fatalf("no objects in the roots")
+		}
+
+		// Create metrics for all Custom Resources in this package.
+		// This creates the customresourcestate.Resource object which contains all metric
+		// definitions for the Custom Resource, if it is part of the package.
+		for _, gv := range kubeKinds {
+			parser.NeedResourceFor(pkg, gv)
+		}
+	}
+
+	// Initialize empty customresourcestate configuration file and fill it with the
+	// customresourcestate.Resource objects from the parser.
+	metrics := customresourcestate.Metrics{
+		Spec: customresourcestate.MetricsSpec{
+			Resources: []customresourcestate.Resource{},
+		},
+	}
+
+	for _, resource := range parser.CustomResourceStates {
+		if resource == nil {
+			continue
+		}
+		if len(resource.Metrics) > 0 {
+			// Sort the metrics to get a deterministic output.
+			sort.Slice(resource.Metrics, func(i, j int) bool {
+				return resource.Metrics[i].Name < resource.Metrics[j].Name
+			})
+
+			metrics.Spec.Resources = append(metrics.Spec.Resources, *resource)
+		}
+	}
+
+	// Sort the resources by GVK to get a deterministic output.
+	sort.Slice(metrics.Spec.Resources, func(i, j int) bool {
+		a := metrics.Spec.Resources[i].GroupVersionKind.String()
+		b := metrics.Spec.Resources[j].GroupVersionKind.String()
+		return a < b
+	})
+
+	// Write the rendered yaml to the context which will result in stdout.
+	filePath := "metrics.yaml"
```

Comment:
Will this overwrite an existing file in the working directory? Can't we store this in a buffer and put this to stdout?

Reply:
This actually will get written to stdout as defined in:

https://github.com/kubernetes/kube-state-metrics/blob/efe2cac20b9d118a7c0daeb72368cbaff2fa9560/pkg/customresourcestate/generate/cmd.go#L82

I'm renaming it tu virtualFilePath to make that more clear.